EIPS/eip-4511.md

```diff
@@ -0,0 +1,116 @@
+
+However, unlike in the EVM, individual BEAM actor-processes are never interrupted during execution of straight-line code. The BEAM "out-of-reductions" check-and-branch is not implemented into the run-loop of the bytecode interpreter itself; but rather, it is a (sometimes-inlined) subroutine implemented into the implementations of two types of instructions: subroutine call/ret instructions, and cross-actor messaging instructions.
+
+Because of this, BEAM bytecode is highly amenable to optimization techniques such as threaded-code interpretation and Just-In-Time compilation. The codegen for individual bytecode instructions can often be as simple as two host-ISA instructions: one to do the required math/stack manip/etc., and the other to subtract an appropriate amount from the reduction counter.
+
+Even better, after codegen, peephole optimization can be applied to hoist all these reduction-counter decrements (which only depend on each-other) out of the basic block they're embedded in, and replace them with a single reduction-counter decrement of the aggregate value from the individual operations, drastically reducing generated code size, and enabling other optimizations due to the code now forming clear "runs" of dependent instructions, without any interruptions from "bookkeeping" code.
+
+### Considerations in Borrowing the BEAM Approach
+
+Note that branching instructions are not included in the set of BEAM instructions that check the reduction counter. This is because the Erlang (and other BEAM language) compilers are careful to never generate branches that represent *back edges* in a Control Flow Graph. Because of this, despite primitive looping constructs being theoretically possible in BEAM bytecode, they're never emitted in practice.
+
+Instead, Erlang and other BEAM languages expose Tail-Call-Optimized (TCOed) recursion as a primitive, and looping constructs are implemented in terms of that primitive. As recursive tail-calls use the "subroutine call" BEAM instruction, these calls always flow through the Out-of-Reductions check-and-branch logic built into that instruction's implementation.
+
+Because of this, BEAM VM languages are free to use *forward* branch instructions to implement non-loop control-structures, e.g. `if/else`, `case`, clause-head pattern matching, `goto fail`-type internal error handling, etc., without paying any bookkeeping overhead cost for these instructions. The instruction implementations for forward-branches remain just as simple as the instruction implementations for math ops *et al*, and so become amenable to optimization (e.g. branchless-code reduction) themselves.
+
+The EVM, as a trustless-multitenant runtime, cannot rely on the developer or compiler kindly limiting themselves to only deploying code that uses a subset of possible instruction sequences, the way BEAM compilers do. The EVM must ensure that Out-of-Gas conditions are checked at least once for each iteration of EVM-primitive branching loops.
+
```

Comment:
Please follow the template, which means Specification comes before Rationale in this case.

https://raw.githubusercontent.com/ethereum/EIPs/master/eip-template.md

Reply:
I'm unclear on what to do in this case, where the Specification refers to information introduced in the Rationale. I originally had them in the opposite order, but the text was more confusing that way.

Reply:
(Also, I would note that I ran `eipv` against this file and it passed; if this is a rule, shouldn't `eipv` enforce it?)

Reply:
The Specification should never depend on any other section.  The specification is the core of the entire document and should be an entirely self contained technical specification.  The specification should not be discussing the *why* of anything, it should merely assert "this is what you do if you want to implement this spec".  Any discussion of the why should happen in the rationale, and that can (and should) reference the specification.

One way to think of it is to write the specification first and imagine that the target audience of this document is some programmer that has been instructed to implement this spec and they don't care at all for the reasoning for anything, they only want to know what the requirements for spec compliance are and they want that information to be as easy to follow as possible.  Any side-tracked discussion about the *why* just makes it harder to implement as it is a distraction from what matters to that person.

Reply:
> Also, I would note that I ran eipv against this file and it passed; if this is a rule, shouldn't eipv enforce it?

Yes, pull requests welcome.  😀